include/libtorrent/aux_/rtc_signaling.hpp

```diff
@@ -0,0 +1,176 @@
+/*
+
+Copyright (c) 2019, Paul-Louis Ageneau
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_RTC_SIGNALING_HPP_INCLUDED
+#define TORRENT_RTC_SIGNALING_HPP_INCLUDED
+
+#include "libtorrent/config.hpp"
+#include "libtorrent/error_code.hpp"
+#include "libtorrent/io.hpp"
+#include "libtorrent/io_context.hpp"
+#include "libtorrent/peer_id.hpp"
+#include "libtorrent/span.hpp"
+#include "libtorrent/time.hpp"
+
+#include <boost/asio/deadline_timer.hpp>
+#include <boost/functional/hash.hpp>
+
+#include <functional>
+#include <memory>
+#include <optional>
+#include <queue>
+#include <vector>
+
+namespace rtc {
+	class PeerConnection;
+	class DataChannel;
+}
+
+namespace libtorrent {
+
+class alert_manager;
+class torrent;
+
+namespace aux {
+
+struct rtc_stream_init;
+
+const int RTC_OFFER_ID_LEN = 16;
+
+struct rtc_offer_id : public std::vector<char> {
```

Comment:
is the `rtc_offer_id` always 16 bytes? If so, I think it would be more appropriate to derive from `std::array<char, RTC_OFFER_ID_LEN>`, or maybe contain one.

The `public` inheritance here is redundant. Inheritance is public by default for `struct`.

It seems a bit awkward to expose all of `std::vector<>`'s interface, including `erase()`, `emplace_back()`, `resize()` etc. Is that really useful for an ID type like this?

Reply:
I don't know if offer ids are supposed to be 16 bytes, I didn't see any spec enforcing it. Only the existing implementation uses 16-bytes ids, so I did the same. I'll look into it to see if it's safe to use a static size here.

On the other points, you're right.